source/Server/IntegratedAuthentication/IntegratedAuthenticationHost.cs

```diff
@@ -0,0 +1,135 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Server.HttpSys;
+using Octopus.Diagnostics;
+using Octopus.Server.Extensibility.Authentication.DirectoryServices.Configuration;
+using Octopus.Server.Extensibility.Authentication.DirectoryServices.DirectoryServices;
+using Octopus.Server.Extensibility.Extensions.Infrastructure.Web;
+using Octopus.Server.Extensibility.HostServices.Web;
+
+namespace Octopus.Server.Extensibility.Authentication.DirectoryServices.IntegratedAuthentication
+{
+    class IntegratedAuthenticationHost : IExecuteWhenServerStarts
+    {
+        readonly ILog log;
+        readonly IWebPortalConfigurationStore configuration;
+        readonly IDirectoryServicesConfigurationStore configurationStore;
+        readonly IIntegratedAuthenticationHandler handler;
+
+        public IntegratedAuthenticationHost(ILog log,
+            IWebPortalConfigurationStore configuration,
+            IDirectoryServicesConfigurationStore configurationStore,
+            IIntegratedAuthenticationHandler handler)
+        {
+            this.log = log;
+            this.configuration = configuration;
+            this.configurationStore = configurationStore;
+            this.handler = handler;
+        }
+
+        public void OnHostStarting() {}
+
+        public void OnHostStarted()
+        {
+            if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                // HttpSys is only supported on Windows, and we're not planning to support Negotiate on Kestrel at this point
+                return;
+            }
+
+            var prefixes = GetListenPrefixes();
+
+            var builder = new WebHostBuilder();
+            
+            builder.UseHttpSys(options =>
+            {
+                options.Authentication.Schemes = MapAuthenticationScheme();
+                options.Authentication.AllowAnonymous = false;
+            
+                foreach (var baseUri in prefixes)
+                {
+                    var prefix = baseUri.ToString();
+            
+                    if (!baseUri.Host.Contains("."))
+                    {
+                        prefix = prefix.Replace("localhost", "+");
+                    }
+            
+                    options.UrlPrefixes.Add(prefix);
+                }
+            });
+            
+             builder.Configure(app =>
+             {
+                 app.Use((context, func) =>
+                 {
+                     if (string.IsNullOrWhiteSpace(context.User.Identity.Name))
+                     {
+                         context.Response.StatusCode = 401;
+                         return Task.CompletedTask;
+                     }
+            
+                     handler.HandleRequest(context);
```

Comment:
I think it would be more idiomatic if the `handler.HandleRequest(context)` was task-returning. This way the handler becomes responsible for deciding how to handle the asynchrony (if any), and your API will not need to change in the future...